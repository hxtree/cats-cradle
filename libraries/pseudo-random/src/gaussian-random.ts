import { ObfuscateCounter } from './obfuscate-counter';

export class GaussianRandom {
  private mean: number;

  private standardDeviation: number;

  private obfuscateCounter;

  constructor(mean: number, standardDeviation: number) {
    this.obfuscateCounter = new ObfuscateCounter();
    this.mean = mean;
    this.standardDeviation = standardDeviation;
  }

  generate(seed: string): number {
    // Using the Box-Muller transform to generate a random number with a Gaussian distribution
    const counter = Number(this.obfuscateCounter.run(seed));
    const max = Number('9'.repeat(seed.length));
    const fraction1 = counter / max;
    const fraction2 = counter / max;

    const u1 = 1 - fraction1;
    const u2 = 1 - fraction2;
    const z0 = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);

    // Apply mean and standard deviation
    return this.mean + z0 * this.standardDeviation;
  }
}

export default GaussianRandom;
