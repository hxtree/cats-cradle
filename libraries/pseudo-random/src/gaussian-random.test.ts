import { GaussianRandom } from './gaussian-random';

describe('GaussianRandom', () => {
  it.each(Array.from({ length: 100 }, (_, index) => index * 9997))(
    'Run %i: should generate numbers with a Gaussian distribution',
    (seed) => {
      const gaussianRandom = new GaussianRandom(0, 1);
      const seedString = seed.toString().padStart(8, '0');

      const result = gaussianRandom.generate(seedString);

      // Snapshot to ensure consistent results across runs
      expect(result).toMatchSnapshot();
    },
  );
});
