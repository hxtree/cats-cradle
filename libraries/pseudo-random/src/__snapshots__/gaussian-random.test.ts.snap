// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GaussianRandom Run 0: should generate numbers with a Gaussian distribution 1`] = `0.03162657460631874`;

exports[`GaussianRandom Run 9997: should generate numbers with a Gaussian distribution 1`] = `0.034645101571436514`;

exports[`GaussianRandom Run 19994: should generate numbers with a Gaussian distribution 1`] = `0.028284889087679135`;

exports[`GaussianRandom Run 29991: should generate numbers with a Gaussian distribution 1`] = `0.019996483252538496`;

exports[`GaussianRandom Run 39988: should generate numbers with a Gaussian distribution 1`] = `0.04472897808016137`;

exports[`GaussianRandom Run 49985: should generate numbers with a Gaussian distribution 1`] = `0.04000374605441871`;

exports[`GaussianRandom Run 59982: should generate numbers with a Gaussian distribution 1`] = `0.024489340286870293`;

exports[`GaussianRandom Run 69979: should generate numbers with a Gaussian distribution 1`] = `0.042430326200861596`;

exports[`GaussianRandom Run 79976: should generate numbers with a Gaussian distribution 1`] = `0.03741634459657112`;

exports[`GaussianRandom Run 89973: should generate numbers with a Gaussian distribution 1`] = `0.014123380275489207`;

exports[`GaussianRandom Run 99970: should generate numbers with a Gaussian distribution 1`] = `0.10495995839094141`;

exports[`GaussianRandom Run 109967: should generate numbers with a Gaussian distribution 1`] = `0.05658249069609298`;

exports[`GaussianRandom Run 119964: should generate numbers with a Gaussian distribution 1`] = `0.05292470324408623`;

exports[`GaussianRandom Run 129961: should generate numbers with a Gaussian distribution 1`] = `0.04899514115109651`;

exports[`GaussianRandom Run 139958: should generate numbers with a Gaussian distribution 1`] = `0.063265566010795`;

exports[`GaussianRandom Run 149955: should generate numbers with a Gaussian distribution 1`] = `0.060015679457143076`;

exports[`GaussianRandom Run 159952: should generate numbers with a Gaussian distribution 1`] = `0.050995655283570786`;

exports[`GaussianRandom Run 169949: should generate numbers with a Gaussian distribution 1`] = `0.061660774062555775`;

exports[`GaussianRandom Run 179946: should generate numbers with a Gaussian distribution 1`] = `0.058321725063539054`;

exports[`GaussianRandom Run 189943: should generate numbers with a Gaussian distribution 1`] = `0.04690378098516911`;

exports[`GaussianRandom Run 199940: should generate numbers with a Gaussian distribution 1`] = `0.08371192025550322`;

exports[`GaussianRandom Run 209937: should generate numbers with a Gaussian distribution 1`] = `0.07213958753265563`;

exports[`GaussianRandom Run 219934: should generate numbers with a Gaussian distribution 1`] = `0.06930623549665248`;

exports[`GaussianRandom Run 229931: should generate numbers with a Gaussian distribution 1`] = `0.06635226672228087`;

exports[`GaussianRandom Run 239928: should generate numbers with a Gaussian distribution 1`] = `0.07749477737039538`;

exports[`GaussianRandom Run 249925: should generate numbers with a Gaussian distribution 1`] = `0.0748639859520992`;

exports[`GaussianRandom Run 259922: should generate numbers with a Gaussian distribution 1`] = `0.06784374581975339`;

exports[`GaussianRandom Run 269919: should generate numbers with a Gaussian distribution 1`] = `0.07618973024858919`;

exports[`GaussianRandom Run 279916: should generate numbers with a Gaussian distribution 1`] = `0.07351233942436129`;

exports[`GaussianRandom Run 289913: should generate numbers with a Gaussian distribution 1`] = `0.06482238447358536`;

exports[`GaussianRandom Run 299910: should generate numbers with a Gaussian distribution 1`] = `0.054773929817554634`;

exports[`GaussianRandom Run 309907: should generate numbers with a Gaussian distribution 1`] = `0.0848966287536716`;

exports[`GaussianRandom Run 319904: should generate numbers with a Gaussian distribution 1`] = `0.08250184414593457`;

exports[`GaussianRandom Run 329901: should generate numbers with a Gaussian distribution 1`] = `0.08003553557602147`;

exports[`GaussianRandom Run 339898: should generate numbers with a Gaussian distribution 1`] = `0.08949266032628708`;

exports[`GaussianRandom Run 349895: should generate numbers with a Gaussian distribution 1`] = `0.08722403569980708`;

exports[`GaussianRandom Run 359892: should generate numbers with a Gaussian distribution 1`] = `0.0812767369357397`;

exports[`GaussianRandom Run 369889: should generate numbers with a Gaussian distribution 1`] = `0.0883647735150886`;

exports[`GaussianRandom Run 379886: should generate numbers with a Gaussian distribution 1`] = `0.08606644571738056`;

exports[`GaussianRandom Run 389883: should generate numbers with a Gaussian distribution 1`] = `0.07877137679698823`;

exports[`GaussianRandom Run 399880: should generate numbers with a Gaussian distribution 1`] = `0.13791479305864532`;

exports[`GaussianRandom Run 409877: should generate numbers with a Gaussian distribution 1`] = `0.09597426265268766`;

exports[`GaussianRandom Run 419874: should generate numbers with a Gaussian distribution 1`] = `0.09386241387703374`;

exports[`GaussianRandom Run 429871: should generate numbers with a Gaussian distribution 1`] = `0.09170192714125797`;

exports[`GaussianRandom Run 439868: should generate numbers with a Gaussian distribution 1`] = `0.10006271639794424`;

exports[`GaussianRandom Run 449865: should generate numbers with a Gaussian distribution 1`] = `0.09803906410488115`;

exports[`GaussianRandom Run 459862: should generate numbers with a Gaussian distribution 1`] = `0.09278732551296276`;

exports[`GaussianRandom Run 469859: should generate numbers with a Gaussian distribution 1`] = `0.09905530723273595`;

exports[`GaussianRandom Run 479856: should generate numbers with a Gaussian distribution 1`] = `0.097010613012597`;

exports[`GaussianRandom Run 489853: should generate numbers with a Gaussian distribution 1`] = `0.09060054177912955`;

exports[`GaussianRandom Run 499850: should generate numbers with a Gaussian distribution 1`] = `0.12255659781840347`;

exports[`GaussianRandom Run 509847: should generate numbers with a Gaussian distribution 1`] = `0.1058985955216898`;

exports[`GaussianRandom Run 519844: should generate numbers with a Gaussian distribution 1`] = `0.10398883407432832`;

exports[`GaussianRandom Run 529841: should generate numbers with a Gaussian distribution 1`] = `0.10204323021246171`;

exports[`GaussianRandom Run 539838: should generate numbers with a Gaussian distribution 1`] = `0.10961660972696523`;

exports[`GaussianRandom Run 549835: should generate numbers with a Gaussian distribution 1`] = `0.10777300802546204`;

exports[`GaussianRandom Run 559832: should generate numbers with a Gaussian distribution 1`] = `0.1030196160910743`;

exports[`GaussianRandom Run 569829: should generate numbers with a Gaussian distribution 1`] = `0.10869804431847278`;

exports[`GaussianRandom Run 579826: should generate numbers with a Gaussian distribution 1`] = `0.10683852194332867`;

exports[`GaussianRandom Run 589823: should generate numbers with a Gaussian distribution 1`] = `0.10105471169598197`;

exports[`GaussianRandom Run 599820: should generate numbers with a Gaussian distribution 1`] = `0.07072071118339561`;

exports[`GaussianRandom Run 609817: should generate numbers with a Gaussian distribution 1`] = `0.11496662740359397`;

exports[`GaussianRandom Run 619814: should generate numbers with a Gaussian distribution 1`] = `0.1132106730380799`;

exports[`GaussianRandom Run 629811: should generate numbers with a Gaussian distribution 1`] = `0.11142686740279231`;

exports[`GaussianRandom Run 639808: should generate numbers with a Gaussian distribution 1`] = `0.11839858698713071`;

exports[`GaussianRandom Run 649805: should generate numbers with a Gaussian distribution 1`] = `0.11669468439692776`;

exports[`GaussianRandom Run 659802: should generate numbers with a Gaussian distribution 1`] = `0.11232139669083467`;

exports[`GaussianRandom Run 669799: should generate numbers with a Gaussian distribution 1`] = `0.11754911089097615`;

exports[`GaussianRandom Run 679796: should generate numbers with a Gaussian distribution 1`] = `0.11583260666774403`;

exports[`GaussianRandom Run 689793: should generate numbers with a Gaussian distribution 1`] = `0.11052260966431629`;

exports[`GaussianRandom Run 699790: should generate numbers with a Gaussian distribution 1`] = `0.13046000134667354`;

exports[`GaussianRandom Run 709787: should generate numbers with a Gaussian distribution 1`] = `0.12336541253105346`;

exports[`GaussianRandom Run 719784: should generate numbers with a Gaussian distribution 1`] = `0.12173174064178631`;

exports[`GaussianRandom Run 729781: should generate numbers with a Gaussian distribution 1`] = `0.12007559550356749`;

exports[`GaussianRandom Run 739778: should generate numbers with a Gaussian distribution 1`] = `0.12656755995528784`;

exports[`GaussianRandom Run 749775: should generate numbers with a Gaussian distribution 1`] = `0.12497627633392426`;

exports[`GaussianRandom Run 759772: should generate numbers with a Gaussian distribution 1`] = `0.12090566434589095`;

exports[`GaussianRandom Run 769769: should generate numbers with a Gaussian distribution 1`] = `0.1257738725087278`;

exports[`GaussianRandom Run 779766: should generate numbers with a Gaussian distribution 1`] = `0.12417228062696214`;

exports[`GaussianRandom Run 789763: should generate numbers with a Gaussian distribution 1`] = `0.11923742290628973`;

exports[`GaussianRandom Run 799760: should generate numbers with a Gaussian distribution 1`] = `0.11408717981311377`;

exports[`GaussianRandom Run 809757: should generate numbers with a Gaussian distribution 1`] = `0.13122171486488285`;

exports[`GaussianRandom Run 819754: should generate numbers with a Gaussian distribution 1`] = `0.1296884163633918`;

exports[`GaussianRandom Run 829751: should generate numbers with a Gaussian distribution 1`] = `0.12813646467135323`;

exports[`GaussianRandom Run 839748: should generate numbers with a Gaussian distribution 1`] = `0.1342337280671147`;

exports[`GaussianRandom Run 849745: should generate numbers with a Gaussian distribution 1`] = `0.13273586263378737`;

exports[`GaussianRandom Run 859742: should generate numbers with a Gaussian distribution 1`] = `0.1289139985224298`;

exports[`GaussianRandom Run 869739: should generate numbers with a Gaussian distribution 1`] = `0.13348637597108143`;

exports[`GaussianRandom Run 879736: should generate numbers with a Gaussian distribution 1`] = `0.13197986373894807`;

exports[`GaussianRandom Run 889733: should generate numbers with a Gaussian distribution 1`] = `0.12735198965176037`;

exports[`GaussianRandom Run 899730: should generate numbers with a Gaussian distribution 1`] = `0.03154104764111958`;

exports[`GaussianRandom Run 909727: should generate numbers with a Gaussian distribution 1`] = `0.13862609019977015`;

exports[`GaussianRandom Run 919724: should generate numbers with a Gaussian distribution 1`] = `0.13717721966914825`;

exports[`GaussianRandom Run 929721: should generate numbers with a Gaussian distribution 1`] = `0.13571252238705322`;

exports[`GaussianRandom Run 939718: should generate numbers with a Gaussian distribution 1`] = `0.14147717881422076`;

exports[`GaussianRandom Run 949715: should generate numbers with a Gaussian distribution 1`] = `0.14005854228615613`;

exports[`GaussianRandom Run 959712: should generate numbers with a Gaussian distribution 1`] = `0.13644611138873566`;

exports[`GaussianRandom Run 969709: should generate numbers with a Gaussian distribution 1`] = `0.14076916325898928`;

exports[`GaussianRandom Run 979706: should generate numbers with a Gaussian distribution 1`] = `0.13934312821732497`;

exports[`GaussianRandom Run 989703: should generate numbers with a Gaussian distribution 1`] = `0.13497286065041955`;
